analytics-core/src/main/java/com/liferay/analytics/android/Analytics.kt

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.analytics.android
+
+import android.app.Application
+import android.content.Context
+import com.liferay.analytics.android.dao.UserDAO
+import com.liferay.analytics.android.model.Event
+import com.liferay.analytics.android.model.Identity
+import com.liferay.analytics.android.model.IdentityContext
+import com.liferay.analytics.android.util.FileStorage
+import com.liferay.analytics.android.util.FlushProcess
+import io.reactivex.annotations.NonNull
+
+/**
+ * @author Igor Matos
+ */
+class Analytics private constructor(fileStorage: FileStorage, internal val analyticsKey: String,
+	flushInterval: Int) {
+
+	private val userDAO: UserDAO
+
+	init {
+		analyticsInstance?.let {
+			throw IllegalStateException("Your library was already initialized.")
+		}
+
+		analyticsInstance = this
+
+		userDAO = UserDAO(fileStorage)
+		flushProcess = FlushProcess(fileStorage, flushInterval)
+	}
+
+	companion object {
+
+		@JvmOverloads
+		@JvmStatic
+		fun configure(@NonNull context: Context, @NonNull analyticsKey: String,
+			flushInterval: Int = FLUSH_INTERVAL_DEFAULT) {
+
+			if (context == null) {
+				throw IllegalArgumentException("Context can't be null.")
+			}
+
+			if (context.applicationContext == null) {
+				throw IllegalArgumentException("AppContext can't be null.")
+			}
+
+			if (analyticsKey.isNullOrEmpty() || analyticsKey.isBlank()) {
+				throw IllegalArgumentException("Analytics key can't be null or empty.")
+			}
+
+			if (flushInterval <= 0) {
+				throw IllegalArgumentException("flushInterval can't be less than or equals zero.")
+			}
+
+			val application = context.applicationContext as Application
+			val fileStorage = FileStorage(application)
+
+			Analytics(fileStorage, analyticsKey, flushInterval)
```

Comment:
Isn't better to have the logic of assigning the instance to the singleton here instead of doing that in the constructor?

Reply:
@victorg1991 We have a cache inside the FlushProcess, so if the flushProcess is in progress we save it on the queue, and when we finish we save it on the files.